### PR TITLE
add compressed double-factorization of T2 amplitudes

### PR DESCRIPTION
**Pull request by WanHsuanLin**

### CODE REVIEW

File: python/ffsim/linalg/double_factorized_decomposition_compressed.py [hunk 1]
```diff
@@ -0,0 +1,353 @@
...
+          axes index the rows and columns of the matrices.
+    """
+    nocc, _, _, _ = t2.shape
+    diag_coulomb_mats, orbital_rotations = double_factorized_t2(t2, tol=tol)
+    _, _, norb, _ = orbital_rotations.shape
+    orbital_rotations = orbital_rotations.reshape(-1, norb, norb)
+    n_reps_full, norb, _ = orbital_rotations.shape
+    diag_coulomb_mats = diag_coulomb_mats.reshape(-1, norb, norb)
+    if n_reps is None or n_reps_full < n_reps:
+        return diag_coulomb_mats, orbital_rotations
```
Comment (kevinsung) [inline python/ffsim/linalg/double_factorized_decomposition_compressed.py]:
```suggestion
    nocc, _, nvrt, _ = t2.shape
    norb = nocc + nvrt

    diag_coulomb_mats, orbital_rotations = double_factorized_t2(t2, tol=tol)
    orbital_rotations = orbital_rotations.reshape(-1, norb, norb)
    diag_coulomb_mats = diag_coulomb_mats.reshape(-1, norb, norb)
    n_reps_full, _, _ = orbital_rotations.shape

    if n_reps is None or n_reps_full < n_reps:
        return diag_coulomb_mats, orbital_rotations

```
simplifies the logic and adds some helpful whitespace.

File: python/ffsim/linalg/double_factorized_decomposition_compressed.py [hunk 2]
```diff
@@ -0,0 +1,353 @@
...
+    if not multi_stage_optimization:
+        n_reps_full = n_reps
+    if begin_reps is None:
+        begin_reps = n_reps_full
+    begin_reps = min(n_reps_full, begin_reps)
```
Comment (kevinsung) [inline python/ffsim/linalg/double_factorized_decomposition_compressed.py]: I found this logic a little hard to follow. I think we make it simpler because we are only doing this to create the `list_reps` variable later. I think we can refactor the logic for creating `list_reps` like this:
```python
if multi_stage_optimization:
    # logic for creating list_reps
else:
    list_reps = [n_reps]

File: python/ffsim/linalg/double_factorized_decomposition_compressed.py [hunk 3]
```diff
@@ -0,0 +1,353 @@
...
+
+    if diag_coulomb_indices is None or len(diag_coulomb_indices) == 0:
```
Comment (kevinsung) [inline python/ffsim/linalg/double_factorized_decomposition_compressed.py]:
```suggestion
    if not diag_coulomb_indices:
```

File: python/ffsim/linalg/double_factorized_decomposition_compressed.py [hunk 4]
```diff
@@ -0,0 +1,353 @@
...
+    *,
+    tol: float = 1e-8,
+    n_reps: int | None = None,
+    diag_coulomb_indices: list[tuple[int, int]] | None = None,
+    method: str = "L-BFGS-B",
+    callback=None,
+    options: dict | None = None,
+    multi_stage_optimization: bool = True,
+    begin_reps: int | None = None,
+    step: int = 2,
```
Comment (kevinsung) [inline python/ffsim/linalg/double_factorized_decomposition_compressed.py]: Let's add an argument `return_optimize_result: bool = False` to allow returning the Scipy OptimizeResult. See https://github.com/qiskit-community/ffsim/blob/914f44b7d00f4ba5c2ca36cbf1c2e4465eabf50e/python/ffsim/variational/orbital_optimization.py#L108-L109

File: python/ffsim/linalg/double_factorized_decomposition_compressed.py [hunk 5]
```diff
@@ -0,0 +1,353 @@
...
+            orbital_rotations_log_jax_tri = orbital_rotations_log_jax * mask
+            n_orb_rot_params = n_tensors * (
+                norb * (norb - 1) // 2 + norb * (norb + 1) // 2
+            )
+            diag_coulomb_params = jnp.array(x[n_orb_rot_params:] + 0j)
+
+            val, (grad_diag_coulomb_params, grad_orbital_rotations_log_jax_tri) = (
+                value_and_grad_func(diag_coulomb_params, orbital_rotations_log_jax_tri)
+            )
+            reshaped_grad = _reshape_grad(
```
Comment (kevinsung) [inline python/ffsim/linalg/double_factorized_decomposition_compressed.py]: I think we can structure the code to make this reshaping unnecessary, by writing a single function that computes the objective function, without having `fun_jax` and `fun_jac` being separate. Any reshaping can be done inside that function. Basically, the structure should be the same as this code: https://github.com/qiskit-community/ffsim/blob/914f44b7d00f4ba5c2ca36cbf1c2e4465eabf50e/python/ffsim/variational/orbital_optimization.py#L139-L178

File: tests/python/linalg/double_factorized_decomposition_compressed_test.py [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+            orbital_rotations,
+            orbital_rotations.conj(),
+        )[:nocc, :nocc, nocc:, nocc:]
+    )
+
+    error = np.sum(np.abs(reconstructed - ccsd.t2) ** 2)
+
+    assert diag_coulomb_mats_optimized.shape == (n_reps, norb, norb)
+    assert orbital_rotations_optimized.shape == (n_reps, norb, norb)
+    assert error_optimized < error
```
Comment (kevinsung) [inline tests/python/linalg/double_factorized_decomposition_compressed_test.py]: Let's make this test more robust by putting a coefficient next to error, i.e. `0.5 * error` (or a smaller factor if it works). Also, once, you add the `return_optimize_result` option, we should add that and bound the numbers of iterations, function calls, and gradient calls. See https://github.com/qiskit-community/ffsim/blob/main/tests/python/variational/orbital_optimization_test.py

File: python/ffsim/linalg/double_factorized_decomposition_compressed.py [hunk 7]
```diff
@@ -0,0 +1,353 @@
...
+        value_and_grad_func = jax.value_and_grad(fun_jax, argnums=(0, 1))
+
+        def fun_jac(x):
+            orbital_rotations_log = _params_to_orb_rot_logs(x, n_tensors, norb)
+            orbital_rotations_log_jax = jnp.array(orbital_rotations_log)
+            mask = jnp.ones((norb, norb), dtype=bool)
+            mask = jnp.triu(mask)
+            orbital_rotations_log_jax_tri = orbital_rotations_log_jax * mask
+            n_orb_rot_params = n_tensors * (
+                norb * (norb - 1) // 2 + norb * (norb + 1) // 2
```
Comment (kevinsung) [inline python/ffsim/linalg/double_factorized_decomposition_compressed.py]: This is just `norb**2`, right?

File: python/ffsim/linalg/double_factorized_decomposition_compressed.py [hunk 8]
```diff
@@ -0,0 +1,385 @@
...
+    eigs, vecs = np.linalg.eigh(-1j * mats)
+    return np.einsum("tij,tj,tkj->tik", vecs, np.exp(1j * eigs), vecs.conj())
+
+
+def _params_to_df_tensors(
+    params: np.ndarray, n_tensors: int, norb: int, diag_coulomb_mat_mask: np.ndarray
+) -> tuple[np.ndarray, np.ndarray]:
+    orb_rot_logs = _params_to_orb_rot_logs(params, n_tensors, norb)
+    orbital_rotations = _expm_antihermitian(orb_rot_logs)
+    n_orb_rot_params = n_tensors * (norb * (norb - 1) // 2 + norb * (norb + 1) // 2)
```
Comment (kevinsung) [inline python/ffsim/linalg/double_factorized_decomposition_compressed.py]:
```suggestion
    n_orb_rot_params = n_tensors * norb**2
```

File: python/ffsim/variational/ucj_spin_balanced.py [hunk 9]
```diff
@@ -429,11 +457,36 @@ def from_t_amplitudes(
 
         nocc, _, nvrt, _ = t2.shape
         norb = nocc + nvrt
+        if optimize:
+            diag_coulomb_indices = []
+            if pairs_aa is not None:
+                diag_coulomb_indices += pairs_aa
+            if pairs_ab is not None:
+                diag_coulomb_indices += pairs_ab
+            diag_coulomb_indices = list(set(diag_coulomb_indices))
```
Comment (kevinsung) [inline python/ffsim/variational/ucj_spin_balanced.py]: If `interaction_pairs` is None, this should end up being None. Instead, it ends up being the empty list.

**Comment by kevinsung** [comment]
TODO
- [ ] document usage in a guide or tutorial
- [ ] consider changing output shape of `linalg.double_factorized_t2` to match the compressed function and merge the two functions
- [ ] add regularization?

File: python/ffsim/linalg/double_factorized_decomposition.py [hunk 10]
```diff
@@ -525,6 +634,106 @@ def double_factorized_t2(
...
+    | tuple[np.ndarray, np.ndarray, scipy.optimize.OptimizeResult]
+):
+    nocc, _, nvrt, _ = t2_amplitudes.shape
+    norb = nocc + nvrt
+
+    diag_coulomb_mats, orbital_rotations = _double_factorized_t2_explicit(
+        t2_amplitudes, tol=tol
+    )
+    orbital_rotations = orbital_rotations.reshape(-1, norb, norb)
+    diag_coulomb_mats = diag_coulomb_mats.reshape(-1, norb, norb)
```
Comment (kevinsung) [inline python/ffsim/linalg/double_factorized_decomposition.py]: Is the reshaping still necessary?
